File: Java Advanced/3. Sets and Maps/Sets and Maps/src/UserLogs.java
import java.util.*;

public class UserLogs {
    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        TreeMap<String, LinkedHashMap<String, Integer>> map = new TreeMap<>();
        String[] line = s.nextLine().split("\\s+");

        while (!"end".equals(line[0])) {
            String[] nameTokens = line[2].split("user=");
            String name = nameTokens[1];

            String[] ipTokens = line[0].split("IP=");
            String ip = ipTokens[1];

            if (!map.containsKey(name)) {
                LinkedHashMap<String, Integer> temp = new LinkedHashMap<>();
                if (!temp.containsKey(ip)) {
                    temp.put(ip, 1);
                } else {
                    temp.put(ip, temp.get(ip) + 1);
                }
                map.put(name, temp);
            } else {
                if (!map.get(name).containsKey(ip)) {
                    map.get(name).put(ip, 1);
                } else {
                    map.get(name).put(ip, map.get(name).get(ip) + 1);
                }
            }

            line = s.nextLine().split("\\s+");
        }
        for (Map.Entry<String, LinkedHashMap<String, Integer>> entry : map.entrySet()) {
            System.out.println(entry.getKey()+":");
            List<String> helper = new LinkedList<>();
            for (Map.Entry<String, Integer> item : entry.getValue().entrySet()) {
                String temp = item.getKey() + " => " + item.getValue();
                helper.add(temp);
            }
            System.out.println(String.join(", ", helper) + ".");
        }
    }
}

File: soft uni prgramming fundamentals/7. Dictionaries, Lambda Expressions and LINQ/Dictionaries/UserLogs/UserLogs.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace UserLogs
{
    class UserLogs
    {
       

        static void Main(string[] args)
        {
            string[] comands = Console.ReadLine().Split(" ").ToArray();
           
            Dictionary<string, Dictionary<string, int>> userInfo = new Dictionary<string, Dictionary<string, int>>();

            while (comands[0] != "end")
            {
                string[] IP = comands[0].Split("=").ToArray();
                string[] user = comands[2].Split("=").ToArray();

                if (userInfo.ContainsKey(user[1]) == false)
                {
                    Dictionary<string, int> temp = new Dictionary<string, int>();

                    if (temp.ContainsKey(IP[1]) == false)
                    {
                        temp.Add(IP[1], 1);
                    }
                    else
                    {
                        temp[IP[1]]++;
                    }
                    userInfo.Add(user[1], temp);
                }
                else
                {
                    if (userInfo[user[1]].ContainsKey(IP[1]) == false)
                    {
                        userInfo[user[1]].Add(IP[1], 1);
                    }
                    else
                    {
                        userInfo[user[1]][IP[1]]++;
                    }
                }
                comands = Console.ReadLine().Split(" ").ToArray();
            }

            foreach (var user in userInfo.OrderBy(x=>x.Key))
            {
                Console.WriteLine($"{user.Key}: ");
                List<string> helper = new List<string>();
                foreach (var item in user.Value)
                {
                    string temp = $"{item.Key} => {item.Value}";
                    helper.Add(temp);
                }
                Console.WriteLine(string.Join(", ",helper)+".");
            }
            {

            }
        }
    }
}
